lib/core/versions/latest/DocumentComposer.ts

```diff
@@ -293,63 +318,52 @@ export default class DocumentComposer {
-    // Find the service of the given service type.
-    let service = undefined;
-    if (document.service !== undefined) {
-      service = document.service.find(service => service.type === patch.serviceType);
+    const serviceEndpoints = patch.serviceEndpoints;
+
+    if (document.service === undefined) {
+      // create a new array if service did not exist
+      document.service = [];
     }
 
-    // If service not found, create a new service element and add it to the property.
-    if (service === undefined) {
-      service = {
-        type: patch.serviceType,
-        serviceEndpoint: {
-          '@context': 'schema.identity.foundation/hub',
-          '@type': 'UserServiceEndpoint',
-          instances: patch.serviceEndpoints
-        }
-      };
+    const idToIndexMapper = new Map();
+    // map all id and their index
+    for (const idx in document.service) {
+      idToIndexMapper.set(document.service[idx].id, idx);
+    }
 
-      if (document.service === undefined) {
-        document.service = [service];
+    for (const serviceEndpoint of serviceEndpoints) {
+      if (idToIndexMapper.has(serviceEndpoint.id)) {
+        const idx = idToIndexMapper.get(serviceEndpoint.id);
+        document.service[idx] = serviceEndpoint;
       } else {
-        document.service.push(service);
-      }
-    } else {
-      // Else we add to the existing service element.
-
-      const serviceEndpointSet = new Set(service.serviceEndpoint.instances);
-
-      // Loop through all given service endpoints and add them if they don't exist already.
-      for (let serviceEndpoint of patch.serviceEndpoints) {
-        if (!serviceEndpointSet.has(serviceEndpoint)) {
-          service.serviceEndpoint.instances.push(serviceEndpoint);
-        }
+        document.service.push(serviceEndpoint);
       }
     }
 
     return document;
   }
 
   private static removeServiceEndpoints (document: DocumentModel, patch: any): DocumentModel {
-    let service = undefined;
-    if (document.service !== undefined) {
-      service = document.service.find(service => service.type === patch.serviceType);
+    if (document.service === undefined) {
+      return document;
     }
 
-    if (service === undefined) {
-      return document;
+    const idToIndexMapper = new Map();
+    // map all id and their index
+    for (const idx in document.service) {
+      idToIndexMapper.set(document.service[idx].id, idx);
     }
 
-    const serviceEndpointSet = new Set(service.serviceEndpoint.instances);
+    const idToRemove = patch.serviceEndpointIds;
 
-    // Loop through all given public key IDs and add them from the existing public key set.
-    for (let serviceEndpoint of patch.serviceEndpoints) {
-      serviceEndpointSet.delete(serviceEndpoint);
+    for (const id of idToRemove) {
+      if (idToIndexMapper.has(id)) {
+        const idx = idToIndexMapper.get(id);
+        document.service.splice(idx, 1);
```

Comment:
Sanity: is this a bug? The fact that splice occurs changes original indices.

Reply:
Yes, it is a bug. Good catch.